### PR TITLE
Implement IsInROI method for Heroes class

### LINKED ISSUE #2271: No sound of vanishing AI hero, after losing a battle
When AI attacks human player's castle and loses a battle, there's no sound, when it disappears from a world map.
image

### PR DESCRIPTION
**Pull request by shprotru**
Code improvement for `Heroes` class, added `Heroes::IsInROI` method

### CODE REVIEW

File: src/fheroes2/heroes/heroes_move.cpp [hunk 1]
```diff
@@ -374,15 +374,24 @@ bool isNeedStayFrontObject( const Heroes & hero, const Maps::Tiles & next )
...
 
-void Heroes::Redraw( fheroes2::Image & dst, s32 dx, s32 dy, bool withShadow ) const
+bool Heroes::IsInROI() const
 {
     const Point & mp = GetCenter();
     const Interface::GameArea & gamearea = Interface::Basic::Get().GetGameArea();
     if ( !( gamearea.GetVisibleTileROI() & mp ) )
+        return false;
+
+    return true;
```
Comment (ihhub) [inline src/fheroes2/heroes/heroes_move.cpp:384]: I think we can do this in one line:
```
return Interface::Basic::Get().GetGameArea().GetVisibleTileROI() & GetCenter();
```

File: src/fheroes2/heroes/heroes.h [hunk 2]
```diff
@@ -279,6 +279,7 @@ class Heroes : public HeroBase, public ColorBase
     void ApplyPenaltyMovement( uint32_t penalty );
     bool ActionSpellCast( const Spell & );
 
+    bool IsInROI() const;
```
Comment (ihhub) [inline src/fheroes2/heroes/heroes.h:282]: This is a helper function which must be `private`. Also please name it from a non-capital letter.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @shprotru could you please take a look into few given comments?

File: src/fheroes2/heroes/heroes.h (hunk 2 above)
Comment (ihhub) [reply src/fheroes2/heroes/heroes.h:282]: Also name it as `isInVisibleMapArea`. ROI is not very descriptive for readers.

File: src/fheroes2/heroes/heroes_move.cpp (hunk 1 above)
Comment (shprotru) [reply src/fheroes2/heroes/heroes_move.cpp]: done

File: src/fheroes2/heroes/heroes.h (hunk 2 above)
Comment (shprotru) [reply src/fheroes2/heroes/heroes.h]: done

**Comment by ihhub** [comment]
Hi @shprotru with all these changes what's the benefit of doing it? From performance point of view we have a slower code.

**Comment by ihhub** [comment]
Hi @shprotru please update the logic of the newly created method to fix #2307.

**Comment by ihhub** [comment]
@shprotru thank you very much for this enhancement!

**Pull request merged**